### PR TITLE
feat(application): add WASM operator for Avro-to-JSON transformation

### LINKED ISSUE #207: feat: Avro-to-JSON WASM Transformation Module
### Problem Statement

Azure IoT Operations (AIO) schema registry does not support Apache Avro or Kafka schema registry formats. When integrating external Kafka sources that publish Avro-encoded messages with AIO's MQTT-based ecosystem, there is no native mechanism to transform these messages into JSON format that downstream consumers expect.

Organizations using industrial IoT systems often have existing Kafka infrastructure producing Avro-serialized telemetry and process control data. Without Avro support in AIO's schema registry, customers must either:

- Modify their existing Kafka producers to emit JSON (breaking existing consumers)
- Deploy separate transformation infrastructure outside AIO (increasing operational complexity)
- Manually decode Avro in each consuming application (duplicating logic)

### Proposed Solution

Create a reusable Rust-based WASM module (`avro-to-json`) that operates as a map operator in Azure IoT Operations dataflow graphs, providing transparent Avro-to-JSON transformation for Kafka-to-MQTT integration scenarios.

### User Story

**As a** solution architect integrating legacy Kafka systems with Azure IoT Operations
**I want** a WASM-based transformation module that converts Avro binary data to JSON
**So that** I can route messages from Kafka topics to AIO MQTT brokers without modifying existing producers or deploying external transformation services

#### Core Functionality

- [ ] WASM module compiles to `wasm32-wasip2` target using Rust toolchain
- [ ] Implements AIO map operator interface via `wasm_graph_sdk` macros
- [ ] Supports multiple Avro encoding formats:
  - Object Container File format (with embedded schema in header)
  - Schema-based decoding using provided Avro schema via configuration
  - Single-object encoding detection (with appropriate logging when schema lookup unavailable)
- [ ] Converts all standard Avro types to JSON equivalents:
  - Primitives (null, boolean, int, long, float, double, string)
  - Complex types (records, arrays, maps, unions, enums)
  - Logical types (dates, timestamps, UUIDs, decimals, durations)
  - Binary data (base64-encoded when not UTF-8)
- [ ] Handles empty payloads gracefully (pass-through with warning)
- [ ] Provides detailed logging at appropriate levels (Info, Warn, Error, Debug)

#### Configuration & Deployment

- [ ] Accepts optional `avroSchema` configuration parameter in JSON format
- [ ] Integrates with blueprint-based deployment pattern:
  - Terraform-based configuration via `full-single-node-cluster` blueprint
  - Example configuration file (`avro-to-json-dataflow.tfvars.example`) in blueprint directory
  - Leverages existing `edge_messaging` module from `src/100-edge/130-messaging`
  - Dataflow graph definition included as part of blueprint variables
- [ ] Provides build scripts that:
  - Compile Rust to WASM (`cargo build --target wasm32-wasip2 --release`)
  - Authenticate with Azure Container Registry
  - Push WASM module as OCI artifact to ACR
  - Push graph definition as separate OCI artifact
- [ ] Blueprint configuration supports:
  - Kafka broker endpoint definition
  - Dataflow graph with Avro-to-JSON transformation node
  - Schema configuration via dataflow properties
  - Registry endpoint reference (ACR integration)

#### Operational Requirements

- [ ] Works within WASM sandbox constraints (no external HTTP calls)
- [ ] Optimized for size and performance (`opt-level = "s"`, LTO enabled)
- [ ] Includes comprehensive unit tests for:
  - Base64 encoding (RFC 4648 compliance)
  - All Avro type conversions
  - Format detection logic
  - Error handling paths

#### Documentation

- [ ] README with:
  - Architecture diagram showing Kafka → AIO → MQTT flow
  - Prerequisites (Rust toolchain, ORAS CLI, Azure CLI, Terraform)
  - Local development quick start (build WASM module)
  - Production deployment via blueprint reference
  - Configuration reference table
  - Integration examples pointing to blueprint tfvars.example
- [ ] Inline code documentation for public APIs
- [ ] Blueprint integration guide referencing `full-single-node-cluster` pattern

#### Component Location

- **Repository Path**: `src/500-application/XXX-avro-to-json/`
- **Component Type**: Edge application workload
- **Deployment Pattern**: WASM module loaded via AIO dataflow graph

#### Integration Points

- **Input**: AIO Kafka source endpoint consuming Avro-encoded messages (configured via `edge_messaging` module)
- **Processing**: Stateless map operator in dataflow graph (WASM module from ACR)
- **Output**: AIO MQTT destination endpoint with JSON-encoded messages
- **Configuration**: Terraform variables defining dataflow graphs with schema provided via properties
- **Blueprint Integration**: Deployed via `full-single-node-cluster` blueprint with custom tfvars configuration
- **Base Infrastructure**: Leverages `src/100-edge/130-messaging` component for dataflow endpoints and graphs

#### Microsoft Learn Documentation

- [AIO Schema Registry Concepts](https://learn.microsoft.com/en-us/azure/iot-operations/connect-to-cloud/concept-schema-registry)
- [Configure Kafka Endpoint](https://learn.microsoft.com/en-us/azure/iot-operations/connect-to-cloud/howto-configure-kafka-endpoint)
- [Develop WASM Modules](https://learn.microsoft.com/en-us/azure/iot-operations/develop-edge-apps/howto-develop-wasm-modules)

#### Repository References

- **Component Base**: `src/500-application/511-rust-embedded-wasm-provider` (WASM module pattern)
- **Messaging Infrastructure**: `src/100-edge/130-messaging` (dataflow endpoints and graphs)
- **Blueprint Pattern**: `blueprints/full-single-node-cluster` (complete deployment example)
- **Example Configurations**:
  - `blueprints/full-single-node-cluster/terraform/dataflow-graphs.tfvars.example`
  - `blueprints/full-single-node-cluster/terraform/dataflow-endpoint.tfvars.example`

#### External Standards

- [Apache Avro Specification](https://avro.apache.org/docs/current/spec.html)
- [WASM Component Model](https://component-model.bytecodealliance.org/)

### PR DESCRIPTION
**Pull request by katriendg**
## Description

This PR introduces component **512-avro-to-json**, a WebAssembly operator that transforms Apache Avro binary data to JSON within Azure IoT Operations dataflow graphs. The implementation targets Kafka-to-MQTT transformation scenarios, providing a stateless, reusable operator with comprehensive type conversion and multiple format detection strategies.

### Core Operator Implementation

The **avro-to-json** operator was implemented with type conversion logic covering all Avro primitives (Boolean, Int, Long, Float, Double, String), complex types (Record, Array, Map, Union, Enum), temporal types (Date, Time, Timestamp), and specialized types (Decimal, Duration, UUID, Bytes, Fixed). The operator includes a custom base64 encoder for binary data without external dependencies, maintaining minimal WASM module size.

Three parsing strategies provide flexibility for different Avro encoding formats:
- **Schema-based parsing**: Uses configured Avro schema for decoding
- **Object Container File detection**: Recognizes OCF format via magic byte verification (`Obj\x01`)
- **Single-object encoding**: Detects the `[C3 01]` marker with schema fingerprint recognition

The operator followed Azure IoT Operations SDK patterns, targeting *wasm32-wasip2*.

### Configuration and Integration

A blueprint example demonstrated complete dataflow graph wiring with **Kafka source endpoint** (SystemAssignedManagedIdentity authentication, TLS enabled), three-node graph topology (Kafka source → WASM map operator → MQTT destination), and embedded equipment telemetry schema. The configuration supported both embedded schemas in OCF format and external schema specification via the *avroSchema* parameter.

### Build and Deployment Automation

Build scripts provided complete automation for WASM compilation (*build-wasm.sh*) with target validation and release profile compilation. Deployment automation (*push-to-acr.sh*) implemented OCI artifact publishing using *oras push*, with proper artifact type declarations (`application/vnd.module.wasm.content.layer.v1+wasm` for modules, `application/vnd.microsoft.aio.graph.v1+yaml` for graph configurations), version extraction from Cargo.toml, and Azure Container Registry authentication.

### Testing and Documentation

Comprehensive test coverage validated all Avro type conversions including edge cases (empty bytes, zero values, negative numbers via two's complement, large decimal values, binary data requiring base64 encoding). Test organization grouped related tests with descriptive names following established patterns.

Component documentation positioned the operator within AIO WASM SDK map operator patterns, outlined the four-step deployment workflow (full stack deployment → WASM target installation → module build → ACR push), documented type mapping rules, and included troubleshooting guidance for build and authentication issues. The deployment workflow emphasized dependencies where infrastructure must be deployed first, creating a pending dataflow graph that resolves asynchronously after artifact publishing.

## Related Issue

Closes #207

## Type of Change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Blueprint modification or addition
- [x] Component modification or addition
- [ ] Documentation update
- [ ] CI/CD pipeline change
- [ ] Other (please describe):

## Implementation Details

The implementation created component **512-avro-to-json** following established patterns from the application layer (500-series). Key technical decisions included:

**Type Conversion Strategy**: Exhaustive pattern matching for all Avro Value variants with custom handling for temporal types (converted to numeric JSON values), Duration (structured objects with months/days/millis fields), and Decimal types (rendered as unscaled integers from big-endian two's complement bytes).

**Format Detection Hierarchy**: Multi-strategy parsing with clear fallback logic (configured schema → OCF format → single-object encoding) and detailed error messages including payload size and hex dumps for diagnostic purposes.

**Dependency Management**: Minimal external dependencies beyond required *apache-avro* and *serde_json* crates. Custom base64 implementation avoided additional dependencies. Used *wasm_graph_sdk* 1.1.3 from Azure IoT Operations SDK preview registry.

**Static Schema Storage**: Thread-safe lazy initialization using `OnceLock<Option<Schema>>` for zero runtime overhead after initialization, appropriate for a stateless transformation operator.

**Version Coordination**: Template variable substitution for VERSION in both graph definitions and module references, enabling consistent versioning during deployment. Documentation emphasized synchronization requirement between Cargo.toml version and Terraform artifact references.

**Build Optimization**: Release profile configured for minimal WASM binary size targeting edge deployment scenarios with constrained resources.

## Testing Performed

- [x] Terraform plan/apply
- [ ] Blueprint deployment test
- [x] Unit tests
- [ ] Integration tests
- [ ] Bug fix includes regression test (see [Test Policy](docs/contributing/testing-validation.md))
- [x] Manual validation
- [ ] Other:

## Checklist

- [x] I have updated the documentation accordingly
- [x] I have added tests to cover my changes
- [x] All new and existing tests passed (25/25 unit tests)
- [ ] I have run `terraform fmt` on all Terraform code (N/A - example file only)
- [ ] I have run `terraform validate` on all Terraform code (N/A - no .tf files)
- [ ] I have run `az bicep format` on all Bicep code (N/A - no Bicep changes)
- [ ] I have run `az bicep build` to validate all Bicep code (N/A - no Bicep changes)
- [x] I have checked for any sensitive data/tokens that should not be committed
- [ ] I have run MegaLinter on my code (`mega-linter-runner`)

## Additional Notes

**Deployment Dependencies**: The four-step deployment workflow requires infrastructure deployment first (creating a pending dataflow graph), followed by WASM target installation, module build, and ACR push. This pattern enables infrastructure-as-code deployment without blocking on WASM compilation.

**Version Synchronization**: Version bumps in Cargo.toml must be synchronized with artifact references in Terraform configuration to ensure graph deployment references the correct WASM module version.

**Schema Flexibility**: The operator supports conditional configuration where the *avroSchema* parameter can be omitted when using Object Container File format with embedded schema, enabling integration with diverse Kafka producer implementations.

### CODE REVIEW

File: blueprints/full-single-node-cluster/terraform/dataflow-graphs-avro-json.tfvars.example
Comment (bindsi) [inline blueprints/full-single-node-cluster/terraform/dataflow-graphs-avro-json.tfvars.example:1]: should we think to move our example tfvars files to a nested folder to have a better structure and visibility for the other important files?

File: src/500-application/512-avro-to-json/operators/avro-to-json/src/lib.rs [hunk 1]
```diff
@@ -0,0 +1,644 @@
...
+                     First 10 bytes (hex): {:02X?}",
+                    payload.len(),
+                    &payload.iter().take(10).copied().collect::<Vec<u8>>()
+                ),
+            })
+        }
+    }
+}
+
+#[cfg(test)]
```
Comment (bindsi) [inline src/500-application/512-avro-to-json/operators/avro-to-json/src/lib.rs:395]: I love the tests!

File: src/500-application/README.md [hunk 2]
```diff
@@ -74,6 +74,8 @@ The following applications are currently available in this directory:
 - **[508-media-connector](./508-media-connector/README.md)** - Akri media connector for camera integration with Azure IoT Operations
 - **[509-sse-connector](./509-sse-connector/README.md)** - Server-Sent Events (SSE) connector for real-time event streaming with Azure IoT Operations
 - **[510-onvif-connector](./510-onvif-connector/README.md)** - ONVIF connector for IP camera integration with Azure IoT Operations
+- **[511-rust-embedded-wasm-provider](./511-rust-embedded-wasm-provider/README.md)** - WebAssembly operators for Azure IoT Operations dataflow graphs using WIT Component Model composition
```
Comment (bindsi) [inline src/500-application/README.md:77]: Thanks for adding this here too :-)

**Review by WilliamBerryiii — APPROVED** [review]

File: blueprints/full-single-node-cluster/terraform/dataflow-graphs-avro-json.tfvars.example
Comment (katriendg) [reply blueprints/full-single-node-cluster/terraform/dataflow-graphs-avro-json.tfvars.example:1]: I think it's a good idea, adding to backlog to tackle in a subsequent PR.

File: src/500-application/512-avro-to-json/operators/avro-to-json/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,644 @@
...
+        }
+        AvroValue::Uuid(u) => json!(u.to_string()),
+        AvroValue::BigDecimal(bd) => {
+            json!(format!("{}", bd))
+        }
+    }
+}
+
+/// Simple base64 encoding without external dependencies
+fn base64_encode(bytes: &[u8]) -> String {
```
Comment (rezatnoMsirhC) [inline src/500-application/512-avro-to-json/operators/avro-to-json/src/lib.rs:108]: Why not use [Crate base64](https://docs.rs/base64/latest/base64/) instead? Might be better than maintaining this complex code and its tests, unless this outweighs using the external dependency.

**Review by rezatnoMsirhC — APPROVED** [review]

File: src/500-application/512-avro-to-json/operators/avro-to-json/src/lib.rs (hunk 3 above)
Comment (katriendg) [reply src/500-application/512-avro-to-json/operators/avro-to-json/src/lib.rs:108]: It's a relevant call-out, we decided to keep it small (WASM output), and because it's only using `decode`, and called only a few times we chose this implemetation.
